0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+
+```
+TODO Embed the definition here
+```
+
+</details>
+
+Settlement engine implementations are RECOMMENDED to implement the standardized endpoints for manual management noted above.
+
+#### Additional RPCs
+
+Settlement engine implementations MAY expose additional, non-standard endpoints for manual operations or configuration.
+
+For example, settlement engines that use payment channels may expose functionality to manually open, fund, and close channels.
+
+### Accounting System HTTP API
+
+#### Credit incoming settlement
+
+Notify the accounting system of receipt of an incoming settlement. Note that the settlement engine MAY accrue incoming settlement acknowledgements without immediately informing the accounting system.
+
+The settlement engine MUST be the only entity invoking this message and MUST credit this amount to the accounts receivable, or some asset account tracking the balance owed by the peer, to ensure accurate accounting.
+
+##### Request
+
+```http
+POST /accounts/:id/receipt HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** that should be credited to the account as an incoming settlement
+
+##### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** credited to the account
+>
+> - If the accounting system uses a unit less precise than the settlement engine's unit in order to prevent the two systems getting out-of-sync, it must indicate the amount it acknowledged receipt of so the settlement engine can track the amount leftover.
+
+### Transport HTTP API
+
+#### Send outgoing request
+
+Send a message to given peer's settlement engine and return its response. The settlement engine MUST be the only entity invoking this instruction.
+
+##### Request
+
+```http
+POST /accounts/:id/sendMessage HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+##### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+### Idempotency
+
+_(The following section was adapted from the [Stripe API documentation](https://stripe.com/docs/api/idempotent_requests).)_
+
+Each API MUST support [idempotency](https://en.wikipedia.org/wiki/Idempotence) for safely retrying requests without accidentally performing the same operation twice. This is useful when an API call is disrupted in transit and you do not receive a response. For example, if a request to send a settlement does not respond due to a network connection error, you can retry the request with the same idempotency key to guarantee that no more than one charge is created.
+
+To perform an idempotent request, provide an additional `Idempotency-Key: <key>` header to the request.
+
+Idempotency works by saving the resulting status code and body of the first request made for any given idempotency key, regardless of whether it succeeded or failed.
```

Comment:
They also need to save the request to the DB before trying to execute it, in case the process crashes after they start executing but before they get to save the intended response